lib/Ast_to_Mir.ml

```diff
@@ -98,21 +95,245 @@ let unquote s =
     String.drop_suffix (String.drop_prefix s 1) 1
   else s
 
+(* hack(sean): strings aren't real
+   XXX add UString to MIR and maybe AST.
+*)
+let mkstring texpr_loc s =
+  { texpr= Lit (Str, s)
+  ; texpr_type= Ast.UReal
+  ; texpr_loc
+  ; texpr_adlevel= DataOnly }
+
 let trans_printables texpr_loc (ps : Ast.typed_expression Ast.printable list) =
   List.map
     ~f:(function
-      | Ast.PString s ->
-          { texpr_adlevel= Ast.DataOnly
-          ; texpr= Lit (Str, unquote s)
-          ; texpr_loc
-          ; texpr_type= UReal }
-          (*XXX hack strings aren't real*)
+      | Ast.PString s -> mkstring texpr_loc (unquote s)
       | Ast.PExpr e -> trans_expr e)
     ps
 
-let rec trans_stmt {Ast.stmt_typed; stmt_typed_loc= sloc; _} =
+(** [add_index expression index] returns an expression that (additionally)
+    indexes into the input [expression] by [index].*)
+let add_int_index e i =
+  let mir_i = trans_idx i in
+  let texpr =
+    match e.texpr with
+    | Var _ -> Indexed (e, [mir_i])
+    | Indexed (e, indices) -> Indexed (e, indices @ [mir_i])
+    | _ -> raise_s [%message "These should go away with Ryan's LHS"]
+  in
+  let texpr_type =
+    Semantic_check.inferred_unsizedtype_of_indexed e.texpr_loc e.texpr_type
+      [(i, UInt)]
+  in
+  {e with texpr; texpr_type}
+
+(** [mkfor] returns a MIR For statement that iterates over the given expression
+    [iteratee]. *)
+let mkfor ut bodyfn iteratee sloc =
```

Comment:
why are you passing in the type separately? why not just use iteratee.texpr_type?

Reply:
this code predates expressions having types. I'll see if I can update it.